### PR TITLE
Initial implementation of meta protocol parser

### PR DESCRIPTION
**Pull request by mrattle**
### TL;DR - What are you trying to accomplish?

This is a refactor of https://github.com/Shopify/async-memcached/pull/35 which pulls the parser out into its own PR for easier review, and fixes errors / uncovered edge cases that were present in the first prototype.

### Details - How are you making this change?  What are the effects of this change?

This PR introduces top-level parsers for Meta Get and Meta Set operations, along with additional relevant helper methods. 
 This change covers all flags that can be returned by a memcached server for all meta operations, allowing for easy implementation of Meta Delete and Meta Arithmetic operations in future PRs.

The strategy here is to build a `MetaValue` struct that contains all the additional (non-value) data requested from the server, which should be piped back to the crate consumer to do as they wish.  This `MetaValue` struct is currently nested into the existing `Value` struct.

This is a functional parsing implementation which will further be refactored to return the `Value` struct back to its original form, leaving `MetaValue` and `Value` as separate concepts.  See: https://github.com/Shopify/async-memcached/pull/64, which will be rebased to cover this new parsing strategy once this is merged.  Usage of these parsers has already been mapped out in the initial prototype PR (https://github.com/Shopify/async-memcached/pull/35), which will be patched up and re-added in a follow-up PR.

### CODE REVIEW

File: src/parser/mod.rs [hunk 1]
```diff
@@ -1,20 +1,65 @@
...
-    pub data: Vec<u8>,
+    pub data: Option<Vec<u8>>,
+    /// optional extra Meta Flags Response Data
+    /// Meta flags are not to be confused with client bitflags, which is an opaque
+    /// number passed by the client. Meta flags change how the command operates, but
+    /// they are not stored in cache.
+    pub meta_values: Option<MetaValue>,
+}
+
+// TODO: Defaults of NONE?
```
Comment (mrattle) [inline src/parser/mod.rs:40]: already done

File: src/parser/mod.rs [hunk 2]
```diff
@@ -30,6 +75,8 @@ pub enum Status {
     Touched,
     /// The key already exists.
     Exists,
+    /// The key already exists and value was requested in the meta protocol.
+    Value,
```
Comment (mrattle) [inline src/parser/mod.rs:79]: This was added because we need a way to differentiate between successful meta get ops where we request the value back with the v flag (e.g. normal get operation) and those where we don't want the value back (e.g. touch to update TTL or a no-op to bump the LRU).

File: .cursorignore
Comment (mrattle) [inline .cursorignore:1]: changes for Cursor codegen

File: src/parser/meta.rs [hunk 3]
```diff
@@ -0,0 +1,813 @@
...
+        alt((
+            value(Response::Status(Status::NotStored), tag(b"NS")),
+            value(Response::Status(Status::Deleted), tag(b"DE")),
+            value(Response::Status(Status::Touched), tag(b"TO")),
+            value(Response::Status(Status::Exists), tag(b"EX")),
+            value(Response::Status(Status::NotFound), tag(b"NF")),
+        )),
+        crlf,
+    )(buf)
+}
```
Comment (mrattle) [inline src/parser/meta.rs:29]: This was an initial approach that will probably be removed in the coming refactor PR, or with the rework of the actual meta_get and meta_set implementations.  It's not possible to use one common status parser because there is overlap / difference in interpretation of status codes across different meta ops.

File: src/parser/meta.rs [hunk 4]
```diff
@@ -0,0 +1,813 @@
...
+        Ok((left, response)) => {
+            let n = bufn - left.len();
+            Ok(Some((n, response)))
+        }
+        Err(nom::Err::Incomplete(_)) => Ok(None),
+        Err(nom::Err::Error(e)) | Err(nom::Err::Failure(e)) => {
+            Err(ErrorKind::Protocol(Some(e.code.description().to_string())))
+        }
+    }
+}
```
Comment (mrattle) [inline src/parser/meta.rs:70]: This will be dropped in favour of the individual op methods above, left here until the completed meta_get and meta_set methods are built off of this PR.

File: src/parser/mod.rs [hunk 5]
```diff
@@ -1,20 +1,65 @@
...
+    pub flags: Option<u32>,
     /// Data for this key.
-    pub data: Vec<u8>,
+    pub data: Option<Vec<u8>>,
+    /// optional extra Meta Flags Response Data
+    /// Meta flags are not to be confused with client bitflags, which is an opaque
+    /// number passed by the client. Meta flags change how the command operates, but
+    /// they are not stored in cache.
+    pub meta_values: Option<MetaValue>,
+}
```
Comment (mrattle) [inline src/parser/mod.rs:38]: Personally not a fan of the nested structure / making everything optional here, but it was easier to leave this as-is for this PR before refactoring so that these are decoupled.

File: src/parser/meta.rs [hunk 6]
```diff
@@ -0,0 +1,813 @@
...
+
+fn map_meta_flag(flag: u8, meta_value: &[u8], value: &mut Value, meta_values: &mut MetaValue) {
+    match flag {
+        b'c' => {
+            value.cas = Some(
+                std::str::from_utf8(meta_value)
+                    .expect("Failed to convert c flag to string")
+                    .parse::<u64>()
+                    .expect("Failed to convert c flag to u64"),
```
Comment (mrattle) [inline src/parser/meta.rs:276]: replace final .expect() with .unwrap_or_else(Error:ParseError)

File: src/parser/meta.rs [hunk 7]
```diff
@@ -0,0 +1,813 @@
...
+        value(Response::Status(Status::Stored), tag(b"HD")),
+        value(Response::Status(Status::NotStored), tag(b"NS")),
+        value(Response::Status(Status::Exists), tag(b"EX")),
+        value(Response::Status(Status::NotFound), tag(b"NF")),
+    ))(buf)
+}
+
+pub fn parse_meta_get_status(buf: &[u8]) -> IResult<&[u8], Response> {
+    alt((
+        value(Response::Status(Status::Value), tag(b"VA ")),
```
[the remark this answers is not in the document]
Comment (mrattle) [reply src/parser/meta.rs:42]: Nuance of now `nom` works.  The ["VA" response](https://github.com/memcached/memcached/blob/5609673ed29db98a377749fab469fe80777de8fd/doc/protocol.txt#L519) (cache hit with value requested / `v` flag used) will always have a space following it so we can parse all 3 characters in one shot in this case, instead of parsing the two character code and then parsing the space to get to the value size.

Example `VA` response:
```
VA <size> <flags>*\r\n.    => VA always followed by whitespace and <size>
<data block>\r\n
```

Alternate response on cache hit with no value requested / `v` flag not used yields `HD` response code which can take these forms:

```
HD <flags>*\r\n    => any non-v flags used, space follows HD

or

HD\r\n     => no flags used, no space following HD, straight into carriage return
```

File: src/parser/meta.rs [hunk 8]
```diff
@@ -0,0 +1,813 @@
...
+        value(Response::Status(Status::Exists), tag(b"HD")),
+        value(Response::Status(Status::NotFound), tag(b"EN")),
+    ))(buf)
+}
+
+// TODO: refactor this into individual parse_meta_*_response methods
+#[allow(dead_code)]
+pub fn parse_meta_response(buf: &[u8]) -> Result<Option<(usize, Response)>, ErrorKind> {
+    let bufn = buf.len();
+    // TODO: alt calls parsers sequentially until one succeeds, which is not optimal.  Want to only call the correct parser, no sequencing.
```
[the remark this answers is not in the document]
Comment (mrattle) [reply src/parser/meta.rs:52]: Yeah, there will be a small amount of code duplication for a _minor_ performance optimization since we won't have to do the fallthrough processing in `nom`'s `alt()` method.

File: src/parser/meta.rs [hunk 9]
```diff
@@ -0,0 +1,813 @@
...
+
+fn parse_meta_flag_values_as_slice(input: &[u8]) -> IResult<&[u8], Vec<(u8, &[u8])>> {
+    if input.is_empty() || input == b"\r\n" {
+        Ok((input, Vec::new()))
+    } else {
+        many0(map(
+            tuple((
+                space1,
+                map(take(1usize), |s: &[u8]| s[0]),
+                take_while1(|c: u8| c != b'\r' && c != b' '), // finding a space means more flags, finding \r means end of flags
```
[the remark this answers is not in the document]
Comment (mrattle) [reply src/parser/meta.rs:262]: In this case we're only parsing the flags, so only need to grab until we hit the `\r` character or a whitespace.  This isn't like a datablock where you can have a mid-way `\r`, `\n` or `\r\n` before hitting your full data size.  The complete `\r\n` that always follows the flags is parsed off with a `crlf()` parser.

File: src/parser/meta.rs [hunk 10]
```diff
@@ -0,0 +1,813 @@
...
+            Ok((input, Response::Data(Some(vec![value]))))
+        }
+        _ => Err(nom::Err::Error(nom::error::Error::new(
+            input,
+            nom::error::ErrorKind::Eof,
+        ))),
+    }
+}
+
+pub fn take_until_size(mut buf: &[u8], byte_size: u32) -> IResult<&[u8], Vec<u8>> {
```
[the remark this answers is not in the document]
Comment (mrattle) [reply src/parser/meta.rs]: Made this change and was able to skip over most of the actual parsing logic in the process, so this is now much cleaner.
